spotify_tensorflow/tfx/tft.py

```diff
@@ -0,0 +1,214 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2017 Spotify AB.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+from __future__ import absolute_import, division, print_function
+
+import argparse
+import os
+import sys
+from abc import abstractmethod, ABCMeta
+from typing import Dict, Union, Any, List  # noqa: F401
+
+import apache_beam as beam
+import six
+from apache_beam.io import tfrecordio
+from apache_beam.io.filesystem import CompressionTypes
+from apache_beam.runners import PipelineState  # noqa: F401
+from spotify_tensorflow.tf_schema_utils import schema_txt_to_feature_spec
+from spotify_tensorflow.tfx.utils import assert_not_empty_string, assert_not_none
+from tensorflow_transform.beam import impl as beam_impl
+from tensorflow_transform.beam.tft_beam_io import transform_fn_io
+from tensorflow_transform.coders import ExampleProtoCoder
+from tensorflow_transform.tf_metadata import dataset_metadata
+from tensorflow_transform.tf_metadata import dataset_schema
+
+
+@six.add_metaclass(ABCMeta)
+class TFTransform:
+
+    """Abstract class for TFX.TFT"""
+
+    @abstractmethod
+    def get_preprocessing_fn(self):  # type: () -> Dict[str, Any]
+        """Returns a users defined preprocessing function"""
+        pass
+
+    def transform_data(self,
+                       pipeline_args,
+                       temp_location,
+                       schema_file,
+                       output_dir,
+                       training_data,
+                       evaluation_data,
+                       transform_fn_dir):
+        tftransform(pipeline_args=pipeline_args,
+                    temp_location=temp_location,
+                    preprocessing_fn=self.get_preprocessing_fn(),
+                    schema_file=schema_file,
+                    output_dir=output_dir,
+                    training_data=training_data,
+                    evaluation_data=evaluation_data,
+                    transform_fn_dir=transform_fn_dir)
+
+    @classmethod
+    def run(cls, args=None):
+        if not issubclass(cls, TFTransform):
+            raise ValueError("Class {} should be inherit from TFT".format(cls))
+
+        parser = argparse.ArgumentParser()
```

Comment:
It seems a bit strange IMO that the user would need to provide these arguments here instead of the luigi class - in fact I'm wondering if it makes sense to move most of this stuff into the luigi class. It would be nice for users to be able to define the transformations and arguments all in one place together with the task definition, wdyt?

Reply:
NVM - discussed this offline, this is only for running outside of luigi. These arguments are needed already in the luigi task.